### PR TITLE
do not call DeleteFile for not-created sst files

### PR DESCRIPTION
**Pull request by jsteemann**
When a memtable is flushed and the flush would lead to a 0 byte .sst
file being created, RocksDB does not write out the empty .sst file to
disk.
However it still calls Env::DeleteFile() on the file as part of some
cleanup procedure at the end of BuildTable().
Because the to-be-deleted file does not exist, this requires
implementors of the DeleteFile() API to check if the file exists on
their own code, or otherwise risk running into PathNotFound errors when
DeleteFile is invoked on non-existing files.
This PR fixes the situation so that when no .sst file is created,
Deletefile will not be called either.
TableFileCreationStarted() will still be called as before.

### CODE REVIEW

**Comment by riversand963** [comment]
iiuc, even if the file does not exist and file deletion fails, the error is ignored. Is this behavior causing trouble for your use case?

```
 Status ignored = fs->DeleteFile(fname, IOOptions(), dbg);
 ignored.PermitUncheckedError();
```

**Comment by jsteemann** [comment]
> iiuc, even if the file does not exist and file deletion fails, the error is ignored. Is this behavior causing trouble for your use case?
> 
> ```
>  Status ignored = fs->DeleteFile(fname, IOOptions(), dbg);
>  ignored.PermitUncheckedError();
> ```

It was first causing light trouble, because we are using a custom `Env` and have overridden `DeleteFile(...)` in it. 
In our override for `DeleteFile(...)` we are dutifully checking the return value of the delete operation and found it to be failing sometimes. And when it failed, we were logging errors.

We first assumed that `DeleteFile(...)` would only be called by RocksDB for actually existing files, but after some analysis we found it to be called also for non-existing files, at least for the case a 0-byte .sst file would be written.

We have worked around the issue in our custom Env's `DeleteFile(...)` override, by running a `IsPathNotFound()` on the the return value of the `DeleteFile(...)` call in case it fails, and then suppress the log message.
This works, but it is just a hack.
IMO it would be more intuitive to not call `DeleteFile(...)` for files that RocksDB does not create in the first place. Otherwise embedders and users of `Env` need to work around that.

**Review by ajkr — APPROVED** [review]
LGTM. Motivation is similar to #8622.

File: db/builder.cc [hunk 1]
```diff
@@ -371,15 +374,19 @@ Status BuildTable(
 
     constexpr IODebugContext* dbg = nullptr;
 
-    Status ignored = fs->DeleteFile(fname, IOOptions(), dbg);
-    ignored.PermitUncheckedError();
+    assert(table_file_created == fs->FileExists(fname, IOOptions(), dbg).ok());
```
Comment (riversand963) [inline db/builder.cc:377]: Is this always guaranteed to be true? What if the file system starts to malfunction just before this `FileExists()` call? Note that file system can be distributed file system and/or NFS, etc.
I think we (try to) use `assert` only to guard against logic errors.

Comment (jsteemann) [reply db/builder.cc]: > Is this always guaranteed to be true? What if the file system starts to malfunction just before this `FileExists()` call? Note that file system can be distributed file system and/or NFS, etc. I think we (try to) use `assert` only to guard against logic errors.

Ok, I have removed the potentially dangerous assert in the latest commit.

**Review by riversand963 — APPROVED** [review]
LGTM

**Pull request closed without merging**